p2p/src/simulated/network.rs

```diff
@@ -482,6 +520,151 @@ impl crate::Receiver for Receiver {
     type Error = Error;
 
     async fn recv(&mut self) -> Result<Message, Error> {
-        self.receiver.next().await.ok_or(Error::PeerClosed)
+        self.receiver.next().await.ok_or(Error::NetworkClosed)
+    }
+}
+
+// A peer in the simulated network.
+// The peer can receive messages over any registered channel.
+struct Peer {
+    // Socket address that the peer is listening on
+    socket: SocketAddr,
+
+    // Map from channel to a sender that can deliver messages to that channel
+    mailboxes: Arc<Mutex<HashMap<Channel, mpsc::UnboundedSender<Message>>>>,
+}
+
+impl Peer {
+    fn new<E: Spawner + RNetwork<Listener, Sink, Stream>>(
+        runtime: &mut E,
+        socket: SocketAddr,
+        max_size: usize,
+    ) -> Self {
+        let result = Self {
+            socket,
+            mailboxes: Arc::new(Mutex::new(HashMap::new())),
+        };
+
+        // Continually listen for incoming connections on the socket
+        runtime.spawn("listener", {
+            let runtime = runtime.clone();
+            let mailboxes = result.mailboxes.clone();
+            async move {
+                let mut listener = runtime.bind(socket).await.unwrap();
+                while let Ok((_, _, mut stream)) = listener.accept().await {
+                    // Spawn a task for this connection
+                    runtime.spawn("receiver", {
+                        let mailboxes = mailboxes.clone();
+                        async move {
+                            // Receive dialer's public key as a handshake
+                            let dialer = match recv_frame(&mut stream, max_size).await {
+                                Ok(data) => data,
+                                Err(_) => {
+                                    error!("failed to receive public key from dialer");
+                                    return;
+                                }
+                            };
+
+                            // Continually receive messages from the dialer and send them to the appropriate mailbox
+                            while let Ok(data) = recv_frame(&mut stream, max_size).await {
+                                let channel = Channel::from_be_bytes(
+                                    data[..SIZE_OF_CHANNEL].try_into().unwrap(),
+                                );
+                                let message = data.slice(SIZE_OF_CHANNEL..);
+                                let mut mailbox = match mailboxes.lock().unwrap().get(&channel) {
+                                    Some(s) => s.clone(),
```

Comment:
nit: is there a way to just get a mutable reference (`get_mut`)? Seems unnecessary to need to clone here.